### PR TITLE
Snake a context through the Chain-blockstore creation

### PR DESCRIPTION
**Pull request by ribasushi**
This sends through the startup daemon context for alternative blockstores that could benefit from it.

## EDIT (by @raulk)

Context: one datastore that will benefit from this is the Postgres-backed datastore. Since it is not an embedded datastore and needs to set up a network connection on initialisation, a context is tremendously useful.

### CODE REVIEW

File: node/repo/interface.go [hunk 1]
```diff
@@ -54,7 +55,7 @@ type LockedRepo interface {
 	Datastore(namespace string) (datastore.Batching, error)
 
 	// Blockstore returns an IPLD blockstore for the requested domain.
```
Comment (raulk) [inline node/repo/interface.go:57]:
```suggestion
	// Blockstore returns an IPLD blockstore for the requested domain.
	// The supplied context must only be used to initialize the blockstore.
	// The implementation should not retain the context for usage throughout
	// the lifecycle.
```

File: node/repo/interface.go [hunk 2]
```diff
@@ -54,7 +55,7 @@ type LockedRepo interface {
 	Datastore(namespace string) (datastore.Batching, error)
```
Comment (raulk) [inline node/repo/interface.go:55]: This looks a bit unlevelled with the Datastore method. They are both backed by some kind of store that would benefit from a cancellatio handle, so we should also propagate a context here.

Comment (ribasushi) [reply node/repo/interface.go:55]: Generally fair, though not within scope of what I am trying to push out at present.
I might circle back to this at a later date when I look into migrating the datastore to be rdbms backed as wel, but I am nowhere near there yet.

@raulk do you consider lack of ctx here blocker for the PR ?

Comment (raulk) [reply node/repo/interface.go:55]: I'm happy to push this to this PR. If we don't do it now, we are piling up one more piece of technical debt IMO.

Comment (ribasushi) [reply node/repo/interface.go:55]: @raulk if you have the bandwidth - by all means, use the same PR

**Review by raulk — CHANGES_REQUESTED** [review]
Signalling blockage until we even out the API, as per review comments. I'll take care of it.

**Review by raulk — APPROVED** [review]
The interfaces are levelled out now. I'm good to merge. Thanks @ribasushi.

**Comment by raulk** [comment]
ping @magik6k for final call / merge.

**Review by raulk — CHANGES_REQUESTED** [review]
Not yet, I forgot to push a commit.

**Comment by raulk** [comment]
Depends on #5425 to fix the build-lotus-soup CI job.

File: node/modules/storageminer.go [hunk 3]
```diff
@@ -357,8 +357,9 @@ func NewProviderPieceStore(lc fx.Lifecycle, ds dtypes.MetadataDS) (dtypes.Provid
 	return ps, nil
 }
 
-func StagingMultiDatastore(lc fx.Lifecycle, r repo.LockedRepo) (dtypes.StagingMultiDstore, error) {
-	ds, err := r.Datastore("/staging")
+func StagingMultiDatastore(lc fx.Lifecycle, mctx helpers.MetricsCtx, r repo.LockedRepo) (dtypes.StagingMultiDstore, error) {
+	ctx := helpers.LifecycleCtx(mctx, lc)
+	ds, err := r.Datastore(ctx, "/staging")
```
Comment (Kubuxu) [inline node/modules/storageminer.go:362]: Doc for the `Datastore` says that ctx is only for startup, here we are passing a lifecycle context from start to the end of Lotus runtime.

Comment (raulk) [reply node/modules/storageminer.go:362]: The `Datastore()` constructor should not retain the context. It should only be used for the start operation, and this is the only context that we have that is cancelled if the startup is cancelled.

Comment (Kubuxu) [reply node/modules/storageminer.go:362]: Hmm, we should be using the context from OnStart here, but it makes everything way harder.

Comment (raulk) [reply node/modules/storageminer.go:362]: Yeah, that would be unnecessarily complex IMO.

**Review by raulk — APPROVED** [review]

**Comment by raulk** [comment]
Approved, but #4995 needs to land first for the tests to pass.

**Review by magik6k — APPROVED** [review]

**Pull request merged**